includes/utils/class-amp-mutation-utils.php

```diff
@@ -0,0 +1,325 @@
+			self::process_markup( $markup );
+			$response['processed_markup'] = $markup;
+		}
+		$response = array_merge( array(
+			self::ERROR_KEY      => self::was_node_removed(),
+			'removed_nodes'      => self::$removed_nodes,
+			'removed_attributes' => self::$removed_attributes,
+		), $response );
+		self::reset_removed();
+
+		return $response;
+	}
+
+	/**
+	 * Reset the stored removed nodes and attributes.
+	 *
+	 * After testing if the markup is valid,
+	 * these static values will remain.
+	 * So reset them in case another test is needed.
+	 *
+	 * @return void.
+	 */
+	public static function reset_removed() {
+		self::$removed_nodes      = null;
+		self::$removed_attributes = null;
+	}
+
+	/**
+	 * Validate the argument in the REST API request.
+	 *
+	 * It would be ideal to simply pass 'is_string' in register_rest_route().
+	 * But it always returned false.
+	 *
+	 * @param mixed $arg The argument to validate.
+	 * @return boolean $is_valid Whether the argument is valid.
+	 */
+	public static function validate_arg( $arg ) {
+		return is_string( $arg );
+	}
+
+	/**
+	 * On updating a post, this checks the AMP validity of the content.
+	 *
+	 * If it's not valid AMP, it adds a query arg to the redirect URL.
+	 * This will cause an error message to appear above the 'Classic' editor.
+	 *
+	 * @param integer $post_id The ID of the updated post.
+	 * @param WP_Post $post The updated post.
+	 * @return void.
+	 */
+	public static function validate_content( $post_id, $post ) {
+		if ( ! self::is_authorized() ) {
+			return;
+		}
+		$filtered_content = apply_filters( 'the_content', $post->post_content );
+		$response         = self::get_response( $filtered_content );
+		if ( isset( $response[ self::ERROR_KEY ] ) && ( true === $response[ self::ERROR_KEY ] ) ) {
+			add_filter( 'redirect_post_location', array( __CLASS__, 'error_message' ) );
+		}
+	}
+
+	/**
+	 * Whether the current user is authorized.
+	 *
+	 * This checks that the user has a certain capability and the nonce is valid.
+	 * It will only return true when updating the post on:
+	 * wp-admin/post.php
+	 * Avoids using check_admin_referer().
+	 * This function might be called in different places,
+	 * and it can't cause it to die() if the nonce is invalid.
+	 *
+	 * @return boolean $is_valid True if the nonce is valid.
+	 */
+	public static function is_authorized() {
+		$nonce = isset( $_REQUEST['_wpnonce'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['_wpnonce'] ) ) : ''; // WPCS: CSRF ok.
+		return ( self::has_cap() && ( false !== wp_verify_nonce( $nonce, 'update-post_' . get_the_ID() ) ) );
+	}
+
+	/**
+	 * Output AMP validation data in the response header of a frontend GET request.
+	 *
+	 * This must be called before the document output begins.
+	 * Because the document is buffered,
+	 * The sanitizers run after the 'send_headers' action.
+	 * So it's not possible to call this function on that hook.
+	 *
+	 * @return void.
+	 */
+	public static function add_header() {
```

Comment:
This isn't actually used yet. I think it was a foundation for validating plugins on activation (#842)